### ISSUE TITLE
PEP 517 Compliant git repos as dependencies, with pipenv, fail install (outdated pip)

### ISSUE THREAD

**Issue by peterhadlaw** [heroku/heroku-buildpack-python#979]
Our team uses `pipenv` and up until now it has been working just fine. We now have to import a private git repository that uses `poetry` internally and thus does not have a `setup.py` defined.

_**I saw in Heroku's blog that `pip` has been upgrade to `20` but is there any process to upgrade the `pipenv` pathway's version of `pip` as well?**_

`pip` `>19` supports this, but the version of `pip` for `pipenv` environments is still back at `9.0.2`

https://github.com/python-poetry/poetry/issues/321
and
https://github.com/pypa/pip/issues/5407

```
-----> Python app detected
-----> Found python-3.6.8, removing
cp: cannot stat '/tmp/build_522b38a153eaf1b98bf6d13aae484a8f/requirements.txt': No such file or directory
-----> Installing python-3.6.10
-----> Installing pip
-----> Installing dependencies with Pipenv 2018.5.18…
       Installing dependencies from Pipfile.lock (bea114)…
       Ignoring appnope: markers 'platform_system == "Darwin"' don't match your environment
       You are using pip version 9.0.2, however version 20.1.1 is available.
       You should consider upgrading via the 'pip install --upgrade pip' command.
       An error occurred while installing git+ssh://git@github.com/<org>/<private_poetry_repo>.git@<git_hash>#egg=<private_poetry_repo>! Will try again.
       Installing initially–failed dependencies…
       Collecting <private_poetry_repo> from git+ssh://git@github.com/<org>/<private_poetry_repo>.git@<git_hash>#egg=<private_poetry_repo>
         Cloning ssh://git@github.com/<org>/<private_poetry_repo>.git (to <git_hash>) to /tmp/pip-build-z1p7odkc/<private_poetry_repo>
       Warning: Permanently added the RSA host key for IP address '140.82.114.3' to the list of known hosts.
           Complete output from command python setup.py egg_info:
           Traceback (most recent call last):
             File "<string>", line 1, in <module>
             File "/app/.heroku/python/lib/python3.6/tokenize.py", line 452, in open
               buffer = _builtin_open(filename, 'rb')
           FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pip-build-z1p7odkc/<private_poetry_repo>/setup.py'
           
           ----------------------------------------
       
         Could not find a tag or branch '<git_hash>', assuming commit.
       Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-z1p7odkc/<private_poetry_repo>/
       You are using pip version 9.0.2, however version 20.1.1 is available.
       You should consider upgrading via the 'pip install --upgrade pip' command.
       
 !     Push rejected, failed to compile Python app.
 !     Push failed
```

**Comment by peterhadlaw** [comment]
FWIW I ended up doing something like this just to get it out, but not sure if you guys are pinning versions or anything else: https://github.com/heroku/heroku-buildpack-python/compare/master...peterhadlaw:master

**Comment by edmorley** [comment]
@peterhadlaw Hi! Thank you for filing this. The current pipenv support has a number of rough edges that are on my list to figure out as new Python owner. We also need to decide long term plans for pipenv support, given the community looks to be wanting to move onto other package managers (eg #921). I likely won't have concrete answers for a while, but will leave this open and update it as I know more.

**Comment by DylanBohlender** [comment]
For what it's worth, there appear to be signs of life from the `pipenv` maintainers - [they've cut a couple of beta releases within the past month](https://github.com/pypa/pipenv/releases) and appear to be working toward a stable release in the near term.

`poetry` support would be excellent too though! Since `pipenv` is the only package manager with first-class support from the Heroku Python buildpack (aside from vanilla `pip` of course), we've avoided straying from `pipenv` but would certainly appreciate more options.

**Comment by edmorley** [comment]
Hi! Sorry for the delay in getting this resolved. I've just shipped #1169 for this, and there's a bit more context in:
https://github.com/heroku/heroku-buildpack-python/issues/987#issuecomment-778151804
